### PR TITLE
fixed terminating stale threads on trap/proc_exit

### PR DESCRIPTION
**Pull request by hritikgupta**
This is to terminate suspended threads in case an atomic wait occurs with a huge or indefinite (-1) timeout, followed by a proc exit or trap.

### CODE REVIEW

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:35]: Why does it have to be a global variable?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:116]: I don't see there's any lock in the `notify_stale_threads` callback. Should we rather have a lock on the `shared_memory_list_lock` mutex at the beginning of the function and release it at the end? Otherwise the map could be modified in the other thread making the traversal corrupted.

Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:116]: the `wasm_runtime_atomic_notify` step in callback calls `acquire_wait_info` which tries to acquire `shared_memory_list_lock`, so I guess encapsulating it with that will cause issues?

Comment (eloparco) [reply core/iwasm/common/wasm_shared_memory.c:116]: `HashMap` doesn't duplicate the key/values pairs, so `wait_map_without_lock` will point to the same elements of `wait_map`

File: core/iwasm/common/wasm_shared_memory.c
Comment (eloparco) [inline core/iwasm/common/wasm_shared_memory.c:124]: we should probably call `bh_hash_map_destroy` on `wait_map_without_lock` at the end

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [reply core/iwasm/common/wasm_shared_memory.c:116]: I think you at least want to lock while you copy hashmaps. Also, what happens if after making a copy a new waiter is added to the map?

Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:116]: While copying the original map, the original map already acquires a lock as it traversal. as inserts happen in the copy map, every insert locks the copy map. 
What could be concerning here is that during the traversal of copy map, the original map could be updated. If I put a lock on the original map during traversal, the execution gets stuck because the callback steps in `notify_stale_threads` try to put a lock on the original map. This might lead to some entries (newly added ones) in the original map not catered to at that point of time, but would be eventually handled, right?

Comment (eloparco) [reply core/iwasm/common/wasm_shared_memory.c:115]: either way I'm still not sure it would be safe since you're not cloning the map but copying the pointers iiuc

**Comment by hritikgupta** [comment]
closing and reopening to reinitiate the build

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:380]: Does the list contain a copy of the list of keys of `wait_map`? If so, why do we need to maintain that list instead of using `wait_map`'s keys?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:114]: should the mutex be initialized?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:112]: does it have to be a linked list? I'd assume we know the exact size of the `wait_map`, so we can use array and only have a single allocation?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:121]: for this traversal, we should probably lock a `shared_memory_list_lock` otherwise the traversal might be invalidated half way through?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [reply core/iwasm/common/wasm_shared_memory.c:114]: I actually don't think we need this mutex at all. What's the risk of calling this function from two different threads? Is it ok if we call `notify` twice on the same address in this specific instance?

Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:114]: I was also skeptical about this, just thought would keep it for the draft and get a second opinion; will remove it.

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:108]: Should we print an error here for debugging purposes?

Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:109]: added

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:106]: I'd suggest allocating `AtomicWaitAddressArgs` on the stack, but dynamically allocate `addr` field (currently it's fixed value - 32), so you don't have to guess the size of the array in advance.

**Review by loganek — CHANGES_REQUESTED** [review]

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:121]: done

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c]: fixed

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [inline core/iwasm/common/wasm_shared_memory.c:134]: @loganek do you think `len`'s value can become inconsistent due to a race condn? do you suggest putting an if check before calling notify i.e. `if(args->addr[i])`?

File: core/iwasm/common/wasm_runtime_common.c
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:2197]: Is this PR to terminating the thread which is atomic waiting, when there is other thread trapping of proc_exiting? Is it duplicated with https://github.com/bytecodealliance/wasm-micro-runtime/pull/1902? Seems #1902 can trigger the signal handler of the atomic waiting thread when os_pthread_kill to it, and in the signal handler, the atomic waiting thread can longjmp into the place of setjmp before, and exit normally?

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:100]: When to free the memory? And why do nothing except LOG_ERROR when malloc/realloc memory failed?

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:129]: Use `/* .. */` to add comments

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:108]: sizeof(void *) * (data->len + 1)?

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:130]: Why not traverse first time to get the total size of wait_map, use it to allocate memory for `AtomicWaitAddressArgs *args`  with size `offsetof(AtomicWaitAddressArgs, addr) + sizeof(void *) * total_size` , and then traverse again without allocating memory?

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:134]: Please declare `uint32 i` at the beginning of function

File: core/iwasm/common/wasm_runtime_common.c
Comment (wenyongh) [reply core/iwasm/common/wasm_runtime_common.c:2197]: Suppose the patch can notify the current atomic wait operation, but if here is another atomic wait operation followed, will the thread be blocked again?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [reply core/iwasm/common/wasm_shared_memory.c:121]: I see you still allocate it on a heap.

File: core/iwasm/common/wasm_runtime_common.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_runtime_common.c:2197]: @wenyongh did you mean #1903?

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:130]: just wondering if there is any specific advantage of traversing two times? is it to ensure the final traversal is safe i.e. we don't go out of bounds?

Comment (loganek) [reply core/iwasm/common/wasm_shared_memory.c:132]: The purpose of that is to avoid multiple reallocations - knowing the size in advance will let you make only one allocation.

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:130]: You don't free the memory allocated for `args` if we reach here (although I still advise you to allocate it on allocate it on the stack instead).

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:126]: Why do you need to add offset of `addr` here?

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:112]:
```suggestion
    AtomicWaitAddressArgs args = { 0 };
```

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:117]:
```suggestion
```

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:142]:
```suggestion
    wasm_runtime_free(args->addr);
```

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c]: that's I think just a compiler extension and won't work with all compilers. I think you can delete it, see my suggestion below.

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:130]:
```suggestion
    args->addr = wasm_runtime_malloc(sizeof(void *) * total_elem_count;
```

**Review by loganek — CHANGES_REQUESTED** [review]

File: core/iwasm/common/wasm_shared_memory.c
Comment (loganek) [inline core/iwasm/common/wasm_shared_memory.c:133]:
```suggestion
    bh_hash_map_traverse(wait_map, create_list_of_waiter_addresses, &args);
```

File: core/iwasm/common/wasm_runtime_common.c
Comment (wenyongh) [reply core/iwasm/common/wasm_runtime_common.c:2197]: Oh, sorry, I meant #1930, but it only works for posix platforms now, maybe this PR is also necessary.

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:122]: Should check the return value, if args.addr is NULL, then unlock shared_memory_list_lock and return?

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:136]: Another issue is the atomic waiting thread might be blocked again after notified by notify_stale_threads_on_exception, when it encounters an atomic_wait opcode again, suggest to check whether there is exception thrown at the beginning of wasm_runtime_atomic_wait in this file, after
```C
    bh_assert(module->module_type == Wasm_Module_Bytecode
              || module->module_type == Wasm_Module_AoT);
```
add
```C
    if (wasm_get_exception(module_inst)) {
        return -1;
    }
```

Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:141]: done

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c]: done

File: core/iwasm/common/wasm_shared_memory.c
Comment (wenyongh) [inline core/iwasm/common/wasm_shared_memory.c:379]: Had better add an empty line below.

Comment (wenyongh) [reply core/iwasm/common/wasm_shared_memory.c:379]: And found this PR is for `dev/wasi_threads`, could you submit to main first? We will merge main into dev/wasi_threads after several PRs are merged.

**Review by loganek — APPROVED** [review]

**Comment by eloparco** [comment]
Does this PR work in AOT mode?

Also, can you try it with the sample [thread_termination.c](https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/wasi_threads/samples/wasi-threads/wasm-apps/thread_termination.c) after replacing https://github.com/bytecodealliance/wasm-micro-runtime/blob/ee1871d3f857a056593ca0d390ce25b6ae492feb/samples/wasi-threads/wasm-apps/thread_termination.c#L34 with a wait operation?

And try with `TEST_TERMINATION_IN_MAIN_THREAD` set to 0 and 1 https://github.com/bytecodealliance/wasm-micro-runtime/blob/ee1871d3f857a056593ca0d390ce25b6ae492feb/samples/wasi-threads/wasm-apps/thread_termination.c#L19

Just to make sure that the base cases are covered.

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:379]: i think changing base to main will cause all commits to dev/wasi_threads added to this PR (since the branch was created from dev/wasi_threads), will create a new PR

Comment (loganek) [reply core/iwasm/common/wasm_shared_memory.c:379]: You can keep this PR, but change base of your `stale_threads_term` branch to be `main` instead of `dev/wasi_threads`.

File: core/iwasm/common/wasm_runtime_common.c
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:2199]: Should move these lines to be after L2204: spread exception firstly to let other threads get exception, and then notify them, so for other thread, if it is atomic waiting and is notified, and if there is another atomic waiting operation, in `wasm_runtime_atomic_wait`, it can detect the exception and return -1 directly. Or it might atomic wait again.

File: core/iwasm/common/wasm_shared_memory.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_shared_memory.c:379]: done

File: core/iwasm/common/wasm_runtime_common.c
Comment (hritikgupta) [reply core/iwasm/common/wasm_runtime_common.c:2203]: makes sense, done

**Pull request merged**
